<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using the package • SparseSignaturesPlus</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using the package">
<meta property="og:description" content="SparseSignaturesPlus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SparseSignaturesPlus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/2_using_the_package.html">Using the package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/danro9685/SparseSignaturesPlus/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="2_using_the_package_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using the package</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/danro9685/SparseSignaturesPlus/blob/HEAD/vignettes/2_using_the_package.rmd" class="external-link"><code>vignettes/2_using_the_package.rmd</code></a></small>
      <div class="hidden name"><code>2_using_the_package.rmd</code></div>

    </div>

    
    
<p>We now present some of the main features of the package. We notice that the package supports different types of mutational signatures such as: SBS (single base substitutions) and MNV (multi-nucleotide variant) (see Degasperi, Andrea, et al. “Substitution mutational signatures in whole-genome–sequenced cancers in the UK population.” Science 376.6591 (2022): abl9283), CX (chromosomal instability) (see Drews, Ruben M., et al. “A pan-cancer compendium of chromosomal instability.” Nature 606.7916 (2022): 976-983) and CN (copy number) signatures (see Steele, Christopher D., et al. “Signatures of copy number alterations in human cancer.” Nature 606.7916 (2022): 984-991). But, for the sake of this vignette, we present only results on the classical SBS signatures. We refer to the manual for details.</p>
<p>First, we show how to load example data and import them into a count matrix to perform the signatures analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/danro9685/SparseSignaturesPlus" class="external-link">"SparseSignaturesPlus"</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: NMF</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: pkgmaker</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: registry</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: rngtools</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: cluster</span></span></code></pre>
<pre><code><span><span class="co">## NMF - BioConductor layer [OK] | Shared memory capabilities [NO: bigmemory] | Cores 2/2</span></span></code></pre>
<pre><code><span><span class="co">##   To enable shared memory capabilities, try: install.extras('</span></span>
<span><span class="co">## NMF</span></span>
<span><span class="co">## ')</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ssm560_reduced</span><span class="op">)</span></span></code></pre></div>
<p>These data are a reduced version of the 560 breast tumors provided by Nik-Zainal, Serena, et al. (2016) comprising only 3 patients. We notice that these data are provided purely as an example, and, as they are a reduced and partial version of the original dataset, they should not be used to draw any biological conclusion.</p>
<p>We now import such data into a count matrix to perform the signatures discovery. To do so, we also need to specify the reference genome as a BSgenome object to be considered. This can be done as follows, where in the example we used hs37d5 as reference genome as provided as data object within the package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.1000genomes.hs37d5"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: BSgenome</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:NMF':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     nrun</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:pkgmaker':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     new2</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biostrings</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: XVector</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biostrings'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     strsplit</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: rtracklayer</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imported_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/get.SBS.counts.html">get.SBS.counts</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ssm560_reduced</span>, reference <span class="op">=</span> <span class="va">BSgenome.Hsapiens.1000genomes.hs37d5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">imported_data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          A[C&gt;A]A A[C&gt;A]C A[C&gt;A]G A[C&gt;A]T A[C&gt;G]A A[C&gt;G]C A[C&gt;G]G A[C&gt;G]T</span></span>
<span><span class="co">## PD10010a      37      25       8      24      35       5      16      25</span></span>
<span><span class="co">## PD10011a     103      59      16      73     113      54      31     102</span></span>
<span><span class="co">## PD10014a     235     241      37     234     158      71      26     180</span></span>
<span><span class="co">##          A[C&gt;T]A A[C&gt;T]C A[C&gt;T]G A[C&gt;T]T A[T&gt;A]A A[T&gt;A]C A[T&gt;A]G A[T&gt;A]T</span></span>
<span><span class="co">## PD10010a      49      31     100      42      21      15      17      30</span></span>
<span><span class="co">## PD10011a     116      73     228     109      61      70      56     165</span></span>
<span><span class="co">## PD10014a     229      89     178     186     105      90     126     174</span></span>
<span><span class="co">##          A[T&gt;C]A A[T&gt;C]C A[T&gt;C]G A[T&gt;C]T A[T&gt;G]A A[T&gt;G]C A[T&gt;G]G A[T&gt;G]T</span></span>
<span><span class="co">## PD10010a      48      20      29      44       8       6      10      23</span></span>
<span><span class="co">## PD10011a     184     116     113     169      77      41      73     105</span></span>
<span><span class="co">## PD10014a     261     122     167     211      76      27      84      59</span></span>
<span><span class="co">##          C[C&gt;A]A C[C&gt;A]C C[C&gt;A]G C[C&gt;A]T C[C&gt;G]A C[C&gt;G]C C[C&gt;G]G C[C&gt;G]T</span></span>
<span><span class="co">## PD10010a      34      28       8      23      15      19      20      26</span></span>
<span><span class="co">## PD10011a     105      75      30     102      60      37      22      65</span></span>
<span><span class="co">## PD10014a     244     238      35     243     107     105      40     144</span></span>
<span><span class="co">##          C[C&gt;T]A C[C&gt;T]C C[C&gt;T]G C[C&gt;T]T C[T&gt;A]A C[T&gt;A]C C[T&gt;A]G C[T&gt;A]T</span></span>
<span><span class="co">## PD10010a      48      37      55      43      12       7      18      16</span></span>
<span><span class="co">## PD10011a      71      52     108     103     116      80      89     103</span></span>
<span><span class="co">## PD10014a     136     124     144     197     116     139     145     217</span></span>
<span><span class="co">##          C[T&gt;C]A C[T&gt;C]C C[T&gt;C]G C[T&gt;C]T C[T&gt;G]A C[T&gt;G]C C[T&gt;G]G C[T&gt;G]T</span></span>
<span><span class="co">## PD10010a      14      17      20      30       6       8       5      13</span></span>
<span><span class="co">## PD10011a     103      78     102     158      40      65      55     188</span></span>
<span><span class="co">## PD10014a     103     144     112     129      47      54      70     107</span></span>
<span><span class="co">##          G[C&gt;A]A G[C&gt;A]C G[C&gt;A]G G[C&gt;A]T G[C&gt;G]A G[C&gt;G]C G[C&gt;G]G G[C&gt;G]T</span></span>
<span><span class="co">## PD10010a      31      22      11      22       6      12       9      14</span></span>
<span><span class="co">## PD10011a      78      50      14      55      55      66      13      87</span></span>
<span><span class="co">## PD10014a     146     126      24     160      63      70      25     120</span></span>
<span><span class="co">##          G[C&gt;T]A G[C&gt;T]C G[C&gt;T]G G[C&gt;T]T G[T&gt;A]A G[T&gt;A]C G[T&gt;A]G G[T&gt;A]T</span></span>
<span><span class="co">## PD10010a      40      32      82      25       6       6       6      13</span></span>
<span><span class="co">## PD10011a      76      63     118      81      69      41      56      86</span></span>
<span><span class="co">## PD10014a     141      99     180     163      62      66      83     126</span></span>
<span><span class="co">##          G[T&gt;C]A G[T&gt;C]C G[T&gt;C]G G[T&gt;C]T G[T&gt;G]A G[T&gt;G]C G[T&gt;G]G G[T&gt;G]T</span></span>
<span><span class="co">## PD10010a      22       9      16      24       7       1       8      10</span></span>
<span><span class="co">## PD10011a      96      62      82      93      56      46      35      99</span></span>
<span><span class="co">## PD10014a     110      81     102     135      32      18      61      78</span></span>
<span><span class="co">##          T[C&gt;A]A T[C&gt;A]C T[C&gt;A]G T[C&gt;A]T T[C&gt;G]A T[C&gt;G]C T[C&gt;G]G T[C&gt;G]T</span></span>
<span><span class="co">## PD10010a      40      40      12      48      54      37      12      85</span></span>
<span><span class="co">## PD10011a      78      80      12      83     116     104      29     194</span></span>
<span><span class="co">## PD10014a     202     191      17     253     198     159      33     325</span></span>
<span><span class="co">##          T[C&gt;T]A T[C&gt;T]C T[C&gt;T]G T[C&gt;T]T T[T&gt;A]A T[T&gt;A]C T[T&gt;A]G T[T&gt;A]T</span></span>
<span><span class="co">## PD10010a      67      55      53      71      39      13       3      35</span></span>
<span><span class="co">## PD10011a     119      94      78     126     121      43      64      91</span></span>
<span><span class="co">## PD10014a     188     153      93     184     124      89      73     221</span></span>
<span><span class="co">##          T[T&gt;C]A T[T&gt;C]C T[T&gt;C]G T[T&gt;C]T T[T&gt;G]A T[T&gt;G]C T[T&gt;G]G T[T&gt;G]T</span></span>
<span><span class="co">## PD10010a      19      13      11      25      18      11      11      35</span></span>
<span><span class="co">## PD10011a     125      79      83     113      68      90     140     251</span></span>
<span><span class="co">## PD10014a     143     118      75     148      71      54      76     160</span></span></code></pre>
<p>Now, we present an example of visualization feature provided by the package, showing the counts for the first patient, i.e., PD10010a, in the following plot.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/patients.SBS.plot.html">patients.SBS.plot</a></span><span class="op">(</span>trinucleotides_counts<span class="op">=</span><span class="va">imported_data</span>,samples<span class="op">=</span><span class="st">"PD10010a"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="2_using_the_package_files/figure-html/unnamed-chunk-3-1.png" alt="Visualization of the counts for patient PD10010a from the dataset published in Nik-Zainal, Serena, et al." width="672"><p class="caption">
Visualization of the counts for patient PD10010a from the dataset published in Nik-Zainal, Serena, et al.
</p>
</div>
<p>After the data are loaded, we can perform signatures de novo extraction. To do so, we need to define a range for the number of signatures (variable K) to be considered. We now show how to perform the inference on the first 3 patients of the dataset from Nik-Zainal, Serena, et al. (2016), whose counts are provided within the package.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">background</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">patients</span><span class="op">)</span></span>
<span><span class="va">res_denovo</span> <span class="op">=</span> <span class="fu"><a href="../reference/signaturesDecomposition.html">signaturesDecomposition</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">patients</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span>, </span>
<span>                                     K <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">4</span>, </span>
<span>                                     background_signature <span class="op">=</span> <span class="va">background</span>, </span>
<span>                                     nmf_runs <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                                     sparsify <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                     num_processes <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                                     seed <span class="op">=</span> <span class="fl">12345</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Performing signatures discovery and rank estimation... </span></span>
<span><span class="co">## Performing inference for K=3... </span></span>
<span><span class="co">## Performing inference for K=4...</span></span></code></pre>
<p>Now that we have performed the de novo inferece, we need to decide the optimal number of signatures to be extracted from our dataset. To do so, we provide a procedure based on cross-validation.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_cv</span> <span class="op">=</span> <span class="fu"><a href="../reference/signaturesCV.html">signaturesCV</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">patients</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span>, </span>
<span>                      beta <span class="op">=</span> <span class="va">res_denovo</span><span class="op">$</span><span class="va">beta</span>, </span>
<span>                      cross_validation_iterations <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                      cross_validation_repetitions <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                      num_processes <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                      seed <span class="op">=</span> <span class="fl">12345</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Estimating the optimal number of signatures with a total of 2 cross validation repetitions... </span></span>
<span><span class="co">## Performing repetition 1 out of 2... </span></span>
<span><span class="co">## Performing estimation for K=3... </span></span>
<span><span class="co">## Performing cross validation iteration 1 out of 2... </span></span>
<span><span class="co">## Performing cross validation iteration 2 out of 2... </span></span>
<span><span class="co">## Progress 50%... </span></span>
<span><span class="co">## Performing estimation for K=4... </span></span>
<span><span class="co">## Performing cross validation iteration 1 out of 2... </span></span>
<span><span class="co">## Performing cross validation iteration 2 out of 2... </span></span>
<span><span class="co">## Progress 100%... </span></span>
<span><span class="co">## Performing repetition 2 out of 2... </span></span>
<span><span class="co">## Performing estimation for K=3... </span></span>
<span><span class="co">## Performing cross validation iteration 1 out of 2... </span></span>
<span><span class="co">## Performing cross validation iteration 2 out of 2... </span></span>
<span><span class="co">## Progress 50%... </span></span>
<span><span class="co">## Performing estimation for K=4... </span></span>
<span><span class="co">## Performing cross validation iteration 1 out of 2... </span></span>
<span><span class="co">## Performing cross validation iteration 2 out of 2... </span></span>
<span><span class="co">## Progress 100%...</span></span></code></pre>
<p>We notice that the computations for this task can be very time consuming, expecially when many iterations of cross validations are performed (see manual) and a large set of configurations of the parameters are tested.</p>
<p>We conclude this vignette by plotting the discovered signatures for the best configuration, i.e., K = 4.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">signatures</span> <span class="op">=</span> <span class="va">res_denovo</span><span class="op">$</span><span class="va">beta</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/signatures.SBS.plot.html">signatures.SBS.plot</a></span><span class="op">(</span>beta<span class="op">=</span><span class="va">signatures</span>, xlabels<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="2_using_the_package_files/figure-html/unnamed-chunk-6-1.png" alt="Visualization of the discovered signatures." width="672"><p class="caption">
Visualization of the discovered signatures.
</p>
</div>
<p>We refer to the manual for a detailed description of each parameter and to the SparseSignaturesPlus manuscript for details on the method.</p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Daniele Ramazzotti, Luca De Sano.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
