<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to the `RESOLVE` R package • RESOLVE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to the `RESOLVE` R package">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RESOLVE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RESOLVE.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/danro9685/RESOLVE/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to the <code>RESOLVE</code> R
package</h1>
                        <h4 data-toc-skip class="author">Luca De
Sano</h4>
                        <h4 data-toc-skip class="author">Daniele
Ramazzotti</h4>
            
            <h4 data-toc-skip class="date">October 15, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/danro9685/RESOLVE/blob/master/vignettes/RESOLVE.rmd" class="external-link"><code>vignettes/RESOLVE.rmd</code></a></small>
      <div class="hidden name"><code>RESOLVE.rmd</code></div>

    </div>

    
    
<p>Cancer is a genetic disease caused by somatic mutations in genes
controlling key biological functions such as cellular growth and
division. Such mutations may arise both through cell-intrinsic and
exogenous processes, generating characteristic mutational patterns over
the genome named mutational signatures. The study of mutational
signatures have become a standard component of modern genomics studies,
since it can reveal which (environmental and endogenous) mutagenic
processes are active in a tumor, and may highlight markers for
therapeutic response.</p>
<p>Mutational signatures computational analyses fall mostly within two
categories: (i) de novo signatures extraction and (ii) signatures
exposure estimation. In the first case, the presence of mutational
processes is first assessed from the data, signatures are identified and
extracted and finally assigned to samples. This task is typically
performed by Non-Negative Matrix Factorization (NMF). While other
approaches have been proposed, NMF-based methods are by far the most
used. The estimation of signatures exposures is performed by holding a
set of signatures fixed (see, e.g., COSMIC mutational signatures
catalogue) and assigning them to samples by minimizing, e.g., mean
squared error between observed and estimated mutational patterns for
each sample.</p>
<p>However, available mutational signatures computational tools presents
many pitfalls. First, the task of determining the number of signatures
is very complex and depends on heuristics. Second, several signatures
have no clear etiology, casting doubt on them being computational
artifacts rather than due to mutagenic processes. Last, approaches for
signatures assignment are greatly influenced by the set of signatures
used for the analysis. To overcome these limitations, we developed
RESOLVE (Robust EStimation Of mutationaL signatures Via rEgularization),
a framework that allows the efficient extraction and assignment of
mutational signatures.</p>
<p>The RESOLVE R package implements a new de novo signatures extraction
algorithm, which employs a regularized Non-Negative Matrix Factorization
procedure. The method incorporates a background signature during the
inference step and adopts elastic net regression to reduce the impact of
overfitting. The estimation of the optimal number of signatures is
performed by bi-cross-validation. Furthermore, the package also provide
a procedure for confidence estimation of signatures activities in
samples.</p>
<p>Additionally, RESOLVE enables the precise stratification of cancer
genomes based on mutational signatures, allowing for the identification
of key signatures that drive specific cancer subtypes. Beyond de novo
extraction, assignment, and confidence estimation, RESOLVE also supports
patient clustering, survival analysis to identify risk groups linked to
distinct mutational processes, and the investigation of associations
between mutational signatures and driver gene mutations. These advanced
capabilities offer deeper insights into tumorigenesis mechanisms, their
prognostic relevance, and potential therapeutic targets.</p>
<p>As such, RESOLVE represents an addition to other Bioconductor
packages, such as, e.g., SparseSignatures, MutationalPatterns, musicatk
among others, that implements a novel approach for detecting mutational
signatures.</p>
<p>In this vignette, we give an overview of the package by presenting
some of its main functions.</p>
<div class="section level2">
<h2 id="installing-the-resolve-r-package">Installing the RESOLVE R package<a class="anchor" aria-label="anchor" href="#installing-the-resolve-r-package"></a>
</h2>
<p>The RESOLVE package can be installed from Bioconductor as follow.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"RESOLVE"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="changelog">Changelog<a class="anchor" aria-label="anchor" href="#changelog"></a>
</h2>
<ul>
<li>
<em>1.0.0</em> - Package released in Bioconductor 3.16.</li>
<li>
<em>1.2.0</em> - Major code refactoring.</li>
<li>
<em>1.10.0</em> - Major code update, implementing signatures-based
clustering and associations to signatures exposures.</li>
</ul>
</div>
<div class="section level2">
<h2 id="using-the-resolve-r-package">Using the RESOLVE R package<a class="anchor" aria-label="anchor" href="#using-the-resolve-r-package"></a>
</h2>
<p>We now present some of the main features of the package. We notice
that the package supports different types of mutational signatures such
as: SBS (single base substitutions) and MNV (multi-nucleotide variant)
(see Degasperi, Andrea, et al. “Substitution mutational signatures in
whole-genome–sequenced cancers in the UK population.” Science 376.6591
(2022): abl9283), CX (chromosomal instability) (see Drews, Ruben M., et
al. “A pan-cancer compendium of chromosomal instability.” Nature
606.7916 (2022): 976-983) and CN (copy number) signatures (see Steele,
Christopher D., et al. “Signatures of copy number alterations in human
cancer.” Nature 606.7916 (2022): 984-991). But, for the sake of this
vignette, we present only results on the classical SBS signatures. We
refer to the manual for details.</p>
<p>First, we show how to load example data and import them into a count
matrix to perform the signatures analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/danro9685/RESOLVE" class="external-link">"RESOLVE"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ssm560_reduced</span><span class="op">)</span></span></code></pre></div>
<p>These data are a reduced version of the 560 breast tumors provided by
Nik-Zainal, Serena, et al. (2016) comprising only 3 patients. We notice
that these data are provided purely as an example, and, as they are a
reduced and partial version of the original dataset, they should not be
used to draw any biological conclusion.</p>
<p>We now import such data into a count matrix to perform the signatures
discovery. To do so, we also need to specify the reference genome as a
BSgenome object to be considered. This can be done as follows, where in
the example we used hs37d5 as reference genome as provided as data
object within the package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.1000genomes.hs37d5"</span><span class="op">)</span></span>
<span><span class="va">imported_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/getSBSCounts.html">getSBSCounts</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ssm560_reduced</span>, reference <span class="op">=</span> <span class="va">BSgenome.Hsapiens.1000genomes.hs37d5</span><span class="op">)</span></span></code></pre></div>
<p>Now, we present an example of visualization feature provided by the
package, showing the counts for the first patient, i.e., PD10010a, in
the following plot.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/patientsSBSPlot.html">patientsSBSPlot</a></span><span class="op">(</span>trinucleotides_counts<span class="op">=</span><span class="va">imported_data</span>,samples<span class="op">=</span><span class="st">"PD10010a"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: `aes_string()` was deprecated in ggplot2 3.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use tidy evaluation idioms with `aes()`.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> See also `vignette("ggplot2-in-packages")` for more information.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">RESOLVE</span> package.</span></span>
<span><span class="co">##   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/danro9685/RESOLVE/issues&gt;</span>.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<div class="figure">
<img src="RESOLVE_files/figure-html/unnamed-chunk-4-1.png" alt="Visualization of the counts for patient PD10010a from the dataset published in Nik-Zainal, Serena, et al." width="672"><p class="caption">
Visualization of the counts for patient PD10010a from the dataset
published in Nik-Zainal, Serena, et al.
</p>
</div>
<p>After the data are loaded, we can perform signatures de novo
extraction. To do so, we need to define a range for the number of
signatures (variable K) to be considered. We now show how to perform the
inference on the dataset from Nik-Zainal, Serena, et al. (2016), whose
counts are provided within the package.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">background</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">patients</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">res_denovo</span> <span class="op">=</span> <span class="fu"><a href="../reference/signaturesDecomposition.html">signaturesDecomposition</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">patients</span>, </span>
<span>                                     K <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, </span>
<span>                                     background_signature <span class="op">=</span> <span class="va">background</span>, </span>
<span>                                     nmf_runs <span class="op">=</span> <span class="fl">100</span>, </span>
<span>                                     num_processes <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>We notice that this function can be also used to perform de novo
estimation for other types of mutational signatures, such as SBS, MNV,
CX and CN.</p>
<p>Now that we have performed the de novo inferece, we need to decide
the optimal number of signatures to be extracted from our dataset. To do
so, we provide a procedure based on cross-validation.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">res_cv</span> <span class="op">=</span> <span class="fu"><a href="../reference/signaturesCV.html">signaturesCV</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">patients</span>, </span>
<span>                      beta <span class="op">=</span> <span class="va">res_denovo</span><span class="op">$</span><span class="va">beta</span>, </span>
<span>                      cross_validation_repetitions <span class="op">=</span> <span class="fl">100</span>, </span>
<span>                      num_processes <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>The computations for this task can be very time consuming, especially
when many iterations of cross validations are performed (see manual) and
a large set of configurations of the parameters are tested.</p>
<p>We refer to the manual for a detailed description of each parameter
and to the RESOLVE manuscript for details on the method.</p>
</div>
<div class="section level2">
<h2 id="signatures-based-clustering-and-associations-to-signatures-exposures">Signatures-based clustering and associations to signatures
exposures<a class="anchor" aria-label="anchor" href="#signatures-based-clustering-and-associations-to-signatures-exposures"></a>
</h2>
<p>RESOLVE also allows researchers to perform downstream analyses once
mutational signatures have been inferred. 1) It enables signatures-based
clustering, using the k-medoids method to group patients with similar
active mutational processes. Additionally, 2) it provides features to
explore associations between signatures and somatic mutations,
identifying links both from signatures to specific mutations and vice
versa. Finally, 3) RESOLVE can examine the relationship between
signature exposures and prognosis, providing insights into the clinical
implications of different mutational processes. By integrating these
analyses, RESOLVE enhances our understanding of the biological and
clinical relevance of mutational signatures.</p>
<p>We next describe how to perform these three analyses with the RESOLVE
R package.</p>
<p>Signatures-based clustering can be performed with the
signaturesClustering functions, which takes as input exposures of
signatures to patients. We suggest to provide normalized exposures.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sbs_assignments</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">norm_alpha</span> <span class="op">=</span> <span class="op">(</span><span class="va">sbs_assignments</span><span class="op">$</span><span class="va">alpha</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">sbs_assignments</span><span class="op">$</span><span class="va">alpha</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sbs_clustering</span> <span class="op">=</span> <span class="fu"><a href="../reference/signaturesClustering.html">signaturesClustering</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="va">norm_alpha</span>, num_clusters <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, num_processes <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Associations of signatures exposures to somatic mutations in driver
genes can be performed either with the associationAlterations function
or the associationSignatures function via regularized regression. In the
first case, mutations are the predictors of the regression and the
signatures exposures the target varibles, which in the second case
regularized logistic regression is performed with the mutations being
the target varibles.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">association_mutations</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">alterations</span> <span class="op">=</span> <span class="va">association_mutations</span><span class="op">$</span><span class="va">alterations</span></span>
<span><span class="va">normalized_alpha</span> <span class="op">=</span> <span class="va">association_mutations</span><span class="op">$</span><span class="va">normalized_alpha</span></span>
<span><span class="va">association_alterations</span> <span class="op">=</span> <span class="fu"><a href="../reference/associationAlterations.html">associationAlterations</a></span><span class="op">(</span>alterations <span class="op">=</span> <span class="va">alterations</span>, signatures <span class="op">=</span> <span class="va">normalized_alpha</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">association_mutations</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">alterations</span> <span class="op">=</span> <span class="va">association_mutations</span><span class="op">$</span><span class="va">alterations</span></span>
<span><span class="va">normalized_alpha</span> <span class="op">=</span> <span class="va">association_mutations</span><span class="op">$</span><span class="va">normalized_alpha</span></span>
<span><span class="va">association_signatures</span> <span class="op">=</span> <span class="fu"><a href="../reference/associationSignatures.html">associationSignatures</a></span><span class="op">(</span>alterations <span class="op">=</span> <span class="va">alterations</span>, signatures <span class="op">=</span> <span class="va">normalized_alpha</span><span class="op">)</span></span></code></pre></div>
<p>Finally, signatures exposures can be associated to prognosis via
regularized Cox regression with the associationPrognosis function. The
results of this analysis are a set of coefficients measuring the impact
of each signatures to prognosis, with positive values representing
positive association with risk (bad prognosis) and negative values
representing negative association with risk (better prognosis).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">association_survival</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">clinical_data</span> <span class="op">=</span> <span class="va">association_survival</span><span class="op">$</span><span class="va">clinical_data</span></span>
<span><span class="va">normalized_alpha</span> <span class="op">=</span> <span class="va">association_survival</span><span class="op">$</span><span class="va">normalized_alpha</span></span>
<span><span class="va">prognosis_associations</span> <span class="op">=</span> <span class="fu"><a href="../reference/associationPrognosis.html">associationPrognosis</a></span><span class="op">(</span>clinical_data <span class="op">=</span> <span class="va">clinical_data</span>, signatures <span class="op">=</span> <span class="va">normalized_alpha</span><span class="op">)</span></span></code></pre></div>
<p>Also in this case, we refer to the manual for a detailed description
of each parameter and to the RESOLVE manuscript for details on the
method.</p>
</div>
<div class="section level2">
<h2 id="current-r-session">Current R Session<a class="anchor" aria-label="anchor" href="#current-r-session"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] BSgenome.Hsapiens.1000genomes.hs37d5_0.99.1</span></span>
<span><span class="co">##  [2] BSgenome_1.76.0                            </span></span>
<span><span class="co">##  [3] rtracklayer_1.68.0                         </span></span>
<span><span class="co">##  [4] BiocIO_1.18.0                              </span></span>
<span><span class="co">##  [5] Biostrings_2.76.0                          </span></span>
<span><span class="co">##  [6] XVector_0.48.0                             </span></span>
<span><span class="co">##  [7] GenomicRanges_1.60.0                       </span></span>
<span><span class="co">##  [8] GenomeInfoDb_1.44.3                        </span></span>
<span><span class="co">##  [9] IRanges_2.42.0                             </span></span>
<span><span class="co">## [10] S4Vectors_0.46.0                           </span></span>
<span><span class="co">## [11] RESOLVE_1.12.0                             </span></span>
<span><span class="co">## [12] Biobase_2.68.0                             </span></span>
<span><span class="co">## [13] BiocGenerics_0.54.1                        </span></span>
<span><span class="co">## [14] generics_0.1.4                             </span></span>
<span><span class="co">## [15] BiocStyle_2.36.0                           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] bitops_1.0-9                gridExtra_2.3              </span></span>
<span><span class="co">##  [3] rlang_1.1.6                 magrittr_2.0.4             </span></span>
<span><span class="co">##  [5] gridBase_0.4-7              matrixStats_1.5.0          </span></span>
<span><span class="co">##  [7] compiler_4.5.1              ggalluvial_0.12.5          </span></span>
<span><span class="co">##  [9] systemfonts_1.3.1           vctrs_0.6.5                </span></span>
<span><span class="co">## [11] reshape2_1.4.4              stringr_1.5.2              </span></span>
<span><span class="co">## [13] pkgconfig_2.0.3             shape_1.4.6.1              </span></span>
<span><span class="co">## [15] crayon_1.5.3                fastmap_1.2.0              </span></span>
<span><span class="co">## [17] labeling_0.4.3              Rsamtools_2.24.1           </span></span>
<span><span class="co">## [19] rmarkdown_2.30              pracma_2.4.4               </span></span>
<span><span class="co">## [21] UCSC.utils_1.4.0            ragg_1.5.0                 </span></span>
<span><span class="co">## [23] xfun_0.53                   glmnet_4.1-10              </span></span>
<span><span class="co">## [25] cachem_1.1.0                jsonlite_2.0.0             </span></span>
<span><span class="co">## [27] SnowballC_0.7.1             DelayedArray_0.34.1        </span></span>
<span><span class="co">## [29] BiocParallel_1.42.2         parallel_4.5.1             </span></span>
<span><span class="co">## [31] cluster_2.1.8.1             R6_2.6.1                   </span></span>
<span><span class="co">## [33] bslib_0.9.0                 stringi_1.8.7              </span></span>
<span><span class="co">## [35] RColorBrewer_1.1-3          jquerylib_0.1.4            </span></span>
<span><span class="co">## [37] Rcpp_1.1.0                  bookdown_0.45              </span></span>
<span><span class="co">## [39] SummarizedExperiment_1.38.1 iterators_1.0.14           </span></span>
<span><span class="co">## [41] knitr_1.50                  Matrix_1.7-4               </span></span>
<span><span class="co">## [43] splines_4.5.1               nnls_1.6                   </span></span>
<span><span class="co">## [45] tidyselect_1.2.1            abind_1.4-8                </span></span>
<span><span class="co">## [47] yaml_2.3.10                 doParallel_1.0.17          </span></span>
<span><span class="co">## [49] codetools_0.2-20            curl_7.0.0                 </span></span>
<span><span class="co">## [51] lattice_0.22-7              tibble_3.3.0               </span></span>
<span><span class="co">## [53] plyr_1.8.9                  withr_3.0.2                </span></span>
<span><span class="co">## [55] S7_0.2.0                    evaluate_1.0.5             </span></span>
<span><span class="co">## [57] desc_1.4.3                  survival_3.8-3             </span></span>
<span><span class="co">## [59] MutationalPatterns_3.18.0   pillar_1.11.1              </span></span>
<span><span class="co">## [61] lsa_0.73.3                  BiocManager_1.30.26        </span></span>
<span><span class="co">## [63] MatrixGenerics_1.20.0       rngtools_1.5.2             </span></span>
<span><span class="co">## [65] foreach_1.5.2               RCurl_1.98-1.17            </span></span>
<span><span class="co">## [67] ggplot2_4.0.0               scales_1.4.0               </span></span>
<span><span class="co">## [69] NMF_0.28                    RhpcBLASctl_0.23-42        </span></span>
<span><span class="co">## [71] glue_1.8.0                  tools_4.5.1                </span></span>
<span><span class="co">## [73] data.table_1.17.8           GenomicAlignments_1.44.0   </span></span>
<span><span class="co">## [75] registry_0.5-1              fs_1.6.6                   </span></span>
<span><span class="co">## [77] XML_3.99-0.19               grid_4.5.1                 </span></span>
<span><span class="co">## [79] colorspace_2.1-2            GenomeInfoDbData_1.2.14    </span></span>
<span><span class="co">## [81] restfulr_0.0.16             cli_3.6.5                  </span></span>
<span><span class="co">## [83] textshaping_1.0.4           S4Arrays_1.8.1             </span></span>
<span><span class="co">## [85] dplyr_1.1.4                 gtable_0.3.6               </span></span>
<span><span class="co">## [87] sass_0.4.10                 digest_0.6.37              </span></span>
<span><span class="co">## [89] SparseArray_1.8.1           rjson_0.2.23               </span></span>
<span><span class="co">## [91] htmlwidgets_1.6.4           farver_2.1.2               </span></span>
<span><span class="co">## [93] htmltools_0.5.8.1           pkgdown_2.1.3.9000         </span></span>
<span><span class="co">## [95] lifecycle_1.0.4             httr_1.4.7</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Daniele Ramazzotti, Luca De Sano.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.9000.</p>
</div>

      </footer>
</div>






  </body>
</html>
